### PR TITLE
Refactor allocate-load API

### PR DESCRIPTION
**Pull request by odow**
Currently just a moving of the allocate-load functions + some documentation. I'll comment in-line about possible changes.

### CODE REVIEW

File: docs/src/submodules/Utilities/overview.md [hunk 1]
```diff
@@ -344,21 +339,29 @@ before the other during a copy, is that the allocate pass needs to create and
...
-If you choose to implement the Allocate-Load API, also implement;
+If you choose to implement the Allocate-Load API, implement the following
+functions, which will be called in order:
+
+**Allocate**
+
+ * [`Utilities.allocate_variables`](@ref)
+ * [`Utilities.allocate_constrained_variable`](@ref)
+ * [`Utilities.allocate_constrained_variables`](@ref)
+ * [`Utilities.allocate`](@ref)
```
Comment (odow) [inline docs/src/submodules/Utilities/overview.md:350]: Should this be `allocate_attribute`?

File: src/Utilities/copy/allocate_load.jl [hunk 2]
```diff
@@ -0,0 +1,445 @@
+# TODO deprecate this in MOI v0.7
+"""
+    needs_allocate_load(model::MOI.ModelLike)::Bool
+
+Return a `Bool` indicating whether `model` does not support `add_variables`/
+`add_constraint`/`set` but supports `allocate_variables`/`allocate_constraint`
+/`allocate`/`load_variables`/`load_constraint`/`load`.
+That is, the allocate-load interface need to be used to copy an model to `model`.
+"""
+needs_allocate_load(::MOI.ModelLike) = false
```
Comment (odow) [inline src/Utilities/copy/allocate_load.jl:10]: I can't find this used anywhere. Is it safe to remove @blegat?

Comment (blegat) [reply src/Utilities/copy/allocate_load.jl:10]: Yes, there is a todo to remove it

**Comment by odow** [comment]
Merging. I'll leave the changes for follow-up PRs.

**Pull request merged**